File: 1. 단계별로_풀어보기/10~19/16. 스택, 큐, 덱 1/4. 4949.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const stringArr = inputArr.slice(0, -1).map(str => str.split(''))

// 출력 결과 변수
let resultStr = ''

// 문장마다 연산
stringArr.forEach(charArr => {
  const stack = Array.from({length: 101}).fill('')
  let stackIdx = 0
  let result = 'yes'

  for (let i = 0; i < charArr.length; i++) {
    const c = charArr[i]

    if (c === '(' || c === '[' || c === '{') {
      stack[stackIdx] = c
      stackIdx += 1
    } //
    else if (c === ')' && stackIdx > 0 && stack[stackIdx - 1] === '(') {
      stackIdx -= 1
      // stack[stackIdx] = '' // 이거 안해도 된다.
    } //
    else if (c === ']' && stackIdx > 0 && stack[stackIdx - 1] === '[') {
      stackIdx -= 1
      // stack[stackIdx] = '' // 이거 안해도 된다
    } //
    else if (c === '}' && stackIdx > 0 && stack[stackIdx - 1] === '{') {
      stackIdx -= 1
      // stack[stackIdx] = '' // 이거 안해도 된다
    } //
    else if (c === ')' || c === ']' || c === '}') {
      result = 'no'
      break
    }
  }

  if (stackIdx > 0) {
    result = 'no'
  }

  resultStr += result + '\n'
})

// 출력
process.stdout.write(resultStr)

File: 1. 단계별로_풀어보기/10~19/16. 스택, 큐, 덱 1/4. 4949.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

while True:
  charArr = list(input())

  if len(charArr) == 2 and charArr[0] == '.':
    break

  result = 'yes'
  stack = deque()

  for c in charArr:
    if c == '(' or c == '[' or c == '{':
      stack.append(c)
    elif c == ')' and len(stack) > 0 and stack[-1] == '(':
      stack.pop()
    elif c == ']' and len(stack) > 0 and stack[-1] == '[':
      stack.pop()
    elif c == '}' and len(stack) > 0 and stack[-1] == '{':
      stack.pop()
    elif c == ')' or c == ']' or c == '}':
      result = 'no'
      break
  
  if len(stack) > 0:
    result = 'no'

  output(f"{result}\n")